ldpd/lde.c

```diff
@@ -1687,6 +1785,278 @@ lde_change_host_label(int af)
+		    != FILTER_PERMIT)
+			/* This neighbor is now filtered so send withdraw */
+			lde_send_labelwithdraw_wcard(ln, NO_LABEL);
+		else {
+			/* This neighbor is allowed in filter list, so
+			 * send labels
+			 */
+			RB_FOREACH(f, fec_tree, &ft) {
+				fn = (struct fec_node *)f;
+				if (fn->fec.type == fec_type &&
+				    fn->local_label != NO_LABEL)
+					lde_send_labelmapping(ln, fn, 0);
+			}
+			lde_imsg_compose_ldpe(IMSG_MAPPING_ADD_END,
+					      ln->peerid, 0, NULL, 0);
+		}
+	}
+}
+
+void
+lde_change_advertise_for_filter(int af)
+{
+	struct lde_nbr  *ln;
+	struct fec      *f;
+	struct fec_node *fn;
+	char            *acl_name;
+	union ldpd_addr *prefix;
+	uint8_t          plen;
+	struct lde_map  *me;
+
+	/* advertise labels for fecs matching this filter */
+	RB_FOREACH(ln, nbr_tree, &lde_nbrs) {
+		RB_FOREACH(f, fec_tree, &ft) {
+			fn = (struct fec_node *)f;
+			switch (af) {
+			case AF_INET:
+				if (fn->fec.type != FEC_TYPE_IPV4)
+					continue;
+				acl_name = ldeconf->ipv4.acl_label_advertise_for;
+				prefix = (union ldpd_addr *)
+					&fn->fec.u.ipv4.prefix;
+				plen = fn->fec.u.ipv4.prefixlen;
+				break;
+			case AF_INET6:
+				if (fn->fec.type != FEC_TYPE_IPV6)
+					continue;
+				acl_name = ldeconf->ipv6.acl_label_advertise_for;
+				prefix = (union ldpd_addr *)
+					&fn->fec.u.ipv6.prefix;
+				plen = fn->fec.u.ipv6.prefixlen;
+				break;
+			default:
+				fatalx("lde_change_egress_label: unknown af");
+			}
+
+			me = (struct lde_map *)fec_find(&ln->sent_map, &fn->fec);
+			if (me && lde_acl_check(acl_name, af, prefix, plen)
+				!= FILTER_PERMIT) {
+				/* This fec is now filtered so send withdraw */
+				lde_send_labelwithdraw(ln, fn, NULL, NULL);
+				lde_map_del(ln, me, 1);
+			}
+			else
+				/* This fec is now allowed in filter, so send
+				 * labels
+				 */
+				if (fn->local_label != NO_LABEL)
+					lde_send_labelmapping(ln, fn, 0);
+		}
+	}
+	RB_FOREACH(ln, nbr_tree, &lde_nbrs)
+		lde_imsg_compose_ldpe(IMSG_MAPPING_ADD_END, ln->peerid, 0,
+			NULL, 0);
+}
+
+void
+lde_change_accept_for_filter(int af)
+{
+	struct lde_nbr  *ln;
+	struct fec      *f;
+	struct fec_node *fn;
+	char            *acl_name;
+	union ldpd_addr *prefix;
+	uint8_t          plen;
+	struct lde_map  *me;
+	struct fec_nh	*fnh;
+	int		 connected = 0;
+
+	/* accept labels from neighbors for fecs defined in this filter */
+	RB_FOREACH(f, fec_tree, &ft) {
+		fn = (struct fec_node *)f;
+		switch (af) {
+		case AF_INET:
+			if (fn->fec.type != FEC_TYPE_IPV4)
+				continue;
+			acl_name = ldeconf->ipv4.acl_label_accept_for;
+			prefix = (union ldpd_addr *)&fn->fec.u.ipv4.prefix;
+			plen = fn->fec.u.ipv4.prefixlen;
+			break;
+		case AF_INET6:
+			if (fn->fec.type != FEC_TYPE_IPV6)
+				continue;
+			acl_name = ldeconf->ipv6.acl_label_accept_for;
+			prefix = (union ldpd_addr *)&fn->fec.u.ipv6.prefix;
+			plen = fn->fec.u.ipv6.prefixlen;
+			break;
+		default:
+			fatalx("lde_change_egress_label: unknown af");
```

Comment:
nit: wrong function name (it's unfortunate `fatalx` doesn't accept format strings so that we could use `__func__` here). Other calls to `fatalx()` need to be updated as well.

Reply:
fixed